연락처 관리앱 [Step2] Rarla, 요한

dpwns1234 opened the pull request:
안녕하세요~ [@LinSaeng](https://github.com/jungseungyeo)~!

연락처 관리 앱 STEP2 PR 보냅니다!
잘 부탁드립니다~

# 프로젝트 구현 요구사항
- UML 작성
- 연락처 목록 화면의 우상단 버튼을 통해 연락처 추가 화면에 진입할 수 있습니다.
- 오토 레이아웃과 스택뷰를 활용하여 화면을 레이아웃합니다.
- 각 필드에 맞는 키보드 종류를 지정합니다.
- 취소 버튼을 선택하면 정말 취소할 것인지 묻도록합니다.
- 저장 버튼을 선택하면 입력한 정보가 올바르게 입력되었는지 확인하도록 합니다.
- 이름은 사용자가 중간에 띄어쓰기를 하더라도 띄어쓰기를 제거합니다.
- 나이는 숫자로만 입력해야하며, 세 자리수 이하여야 합니다.
- 연락처는 중간에 -로 구분하며, -는 두 개 존재해야 합니다.
- -을 제외하고 숫자는 9자리 이상이어야 합니다.
- 올바른 사용자 입력 예시

|이름|나이|연락처|
|------|---|---|
|james|30|05-343-2443|
|ja mes|30|013-3343-2343|
|james|3 0|013-3432-3433|

- 올바르지 않은 사용자 입력 예시

|이름|나이|연락처|
|------|---|---|
||999999999|30|
||-1|013-432-235|
||01033532354|-|

- 모든 정보가 올바르게 입력되었다면 저장 버튼을 선택했을 때 이전 화면으로 돌아가고, 추가한 연락처가 연락처 리스트에 추가됩니다.
<br/>

# UML
<image src="https://hackmd.io/_uploads/H1vgWYUbT.png"/>
<br/><br/><br/>


# 코드를 작성하며 고민한 점
1. 유효성을 검사하는 코드를 작성할 때 TDD 형태로 구현해보았습니다.
2. 연락처 데이터는 DTO로 순수하게 데이터만을 관리하고자 Contact Model을 분리했습니다.
3. Contact Model을 위한 Logic을 관리하기 위한 코드는 ContactManager 내부에서 구현했습니다.
4. ContactManagerViewController에서 AddContactViewController에 manager 전달 시 의존성 주입 형태로 전달하고자 했습니다.
5. 연락처 추가 화면에서 저장 버튼 클릭 시 연락처 리스트 화면에서 변화를 인지하고 테이블을 다시 그리게 하고자 옵져버 패턴인 Notification를 활용해 구현했습니다.
6. Alert 생성 시 중복되는 코드를 줄이고자 AlertViewController를 만들어 처리했습니다.
7. AlertViewController를 호출 시 함수 형태를 전달함에 있어 excaping 키워드와 클로저에 대한 추가적인 학습을 진행했습니다.
8. 연락처 입력 시 -를 자동으로 입력해주는 처리가 필요해 로직을 찾아 구현했습니다.
<br/>

# 프로젝트 하면서 궁금한 점
- UML이 잘 작성되었다고 생각하시는지 궁금합니다! 잘못된 점이나 보충해야할 부분이 있을까요~?
- 고민한 점 4번에서 manager를 전달할 때 Singleton 패턴은 이전 프로젝트로 구현해 본 경험이 있어 다른 방식을 찾던 중 의존성 주입을 하는 형태를 알게되어 해당 방식으로 구현했습니다. 작성한 코드가 의존성 주입을 올바르게 처리한 방법인지 명확하지 않아 확인 부탁드립니다!
- 중복되는 코드를 줄이고자 AlertViewController를 만들어 구현했습니다. 이전 PR에서 동일한 문제 해결방식을 물어보았는데 Factory pattern을 활용해보라는 답변을 받았다고 해서 고민해보았습니다. 
    - 실무에서는 팝업을 구현할 때 Factory pattern을 활용해 구현하는지 궁금합니다.
    - Factory pattern을 구현하는 형태를 [링크1](https://icksw.tistory.com/237), [링크2](https://betterprogramming.pub/a-better-way-to-implement-alerts-on-your-swift-ios-projects-70de896bcdc2)를 참고해 고민해보았습니다. 하지만 중복되는 코드를 줄인다고 생각되지 않고 불필요한 메서드도 전달해줘야하는 형태로 생각되어 지금의 형태로 구현했습니다. 
    - 현재의 형태도 하나의 버튼의 Alert, 두개의 버튼의 Alert를 만들때 명확히 전달해주는 방식이 아니라 개선이 필요해보입니다. 더 나은 방식이 있을지 조언 부탁드립니다!
- AlertViewController의 showAlert()메서드를 호출할 때 너무 많은 매개변수를 받는 것 같아, AlertData 구조체를 만들어 showAlert(AlertData)로 전달하도록 구현했습니다. 이렇게 리팩터하는 것이 맞는지, 매개변수의 개수를 몇 개를 넘어서면 안되는지에 대한 기준이 있는지 궁금합니다.

dpwns1234 committed:
refactor: ContactManagerVC의 캡슐화

rarlala committed:
feat: AddContactView 스토리보드 화면 추가

dpwns1234 committed:
design: 연락처 추가 화면 추가

rarlala committed:
feat: TDD 방식으로 validate check 함수 구현

dpwns1234 committed:
feat: 연락처 추가 화면 이동 구현과 연락처 저장시 유효성 검사

rarlala committed:
feat: Error Model 정의 및 에러에 맞게 Alert 출력

dpwns1234 committed:
feat: 연락처 추가

- 제대로 작동 안함.

dpwns1234 committed:
feat: NotificationCenter를 활용한 tableView data update

전 커밋의 문제 해결
ContactManager의 type을 struct -> class 로 변경으로 해결

rarlala committed:
feat: cancel 버튼 클릭 시 안내 팝업 추가

dpwns1234 committed:
refactor: validate 관련 함수 리팩터

- validate 함수 안에서 가능한 에러 처리를 모두 넣어줌

rarlala committed:
refactor: 저장 버튼 클릭 시 에러 메세지 출력 함수 refactoring

rarlala committed:
refactor: AlearController 추가 후 로직 수정

dpwns1234 committed:
feat: 연락처 입력시 자동 하이픈 적용

dpwns1234 committed:
refactor: naming 변경 및 적절한 접근제어자 설정

dpwns1234 committed:
refactor: showAlert()메서드의 매개변수를 하나의 구조체로 합쳐서 보내도록 리팩터

dpwns1234 committed:
refactor: AlertViewController의 인스턴스가 아닌 변수를 만들어 사용

dpwns1234 committed:
fix: phone validate 논리 오류 수정

rarlala committed:
fix: formatingPhone 중복 로직 제거 및 코드 수정, validate 공백 제거 누락 코드 추가

jungseungyeo commented on ContactManager/ContactManager/Controller/AddContactViewController.swift:
```diff
@@ -0,0 +1,88 @@
+
+// MARK: - extension
+
+extension String {
```
별도의 폴더를 만들어서 적용하면 좋을거같아요. 이런 기능 확장은 여러군데에서 사용이 가능할수있어요

jungseungyeo replied:
별도의 파일로 적용하면 접근성이 더 용이 할거같아요.

rarlala replied:
넵! [941f397](https://github.com/tasty-code/ios-contact-manager-ui/pull/34/commits/941f397ab8402f55e3bf0e466c2f9d5a892fb3c2) 에서 수정했습니다 :)

jungseungyeo commented on ContactManager/ContactManager/Controller/AddContactViewController.swift:
```diff
@@ -0,0 +1,88 @@
+
+extension String {
+  func formatingPhone() -> String {
+    let _str = self.replacingOccurrences(of: "-", with: "")
```
_ <- 네이밍을 swift에서는 잘 사용하지 않는 네이밍입니다. 큰 문제는 없지만 일반적으로 objective-c를 사용했던 개발자나 java, 안드로이드 출신 개발자분에서 많이 보이는데요. 
swift 언어를 잘 알고있는 느낌을 주기 위해서는 _ <-로 시작하는 네이밍을 지양하면 좋아요.

제가 좋아하는 스타일쉐어 스타일 가이드입니다. 참고해주세요 
https://github.com/StyleShare/swift-style-guide#%EB%84%A4%EC%9D%B4%EB%B0%8D

rarlala replied:
넵! [941f397](https://github.com/tasty-code/ios-contact-manager-ui/pull/34/commits/941f397ab8402f55e3bf0e466c2f9d5a892fb3c2) 에서 수정했습니다 :)

jungseungyeo commented on ContactManager/ContactManager/Controller/AlertViewController.swift:
```diff
@@ -0,0 +1,28 @@
+    super.viewDidLoad()
+  }
+  
+  func showAlert(_ data: AlertData) {
```
일회용으로 사용하는 alert용 ViewController이네요. 

이거 static func으로 만들어서 사용하는것은 어떤가요?

jungseungyeo replied:
static 함수로 이루어진 객체이면 재사용성이 많이 올라갈거같아요.

dpwns1234 replied:
감사합니다! 타입 메서드에 대해 알아갈 수 있었습니다. 변경사항은 [d456bd9](https://github.com/tasty-code/ios-contact-manager-ui/pull/34/commits/77fd2392458e0b9e31ae95bf4316bedc5d456bd9) 커밋에서 수정하였습니다!

jungseungyeo commented on ContactManager/ContactManager/Model/AlertData.swift:
```diff
@@ -0,0 +1,16 @@
+  var message: String
+  var defaultButtonTitle: String
+  var destructiveButtonTitle: String? = nil
+  var destructiveAction: (() -> Void)? = nil
```
변수로 작성된 이유가 있나요? 
또 이렇게 alertData보다는 AlertViewController의 의존성을 넣어주면 어디서 사용하는지 명확히 알수있을거같아요.

`extension AlertViewController  {
   struct AlearDtata { ...
}
`

dpwns1234 replied:
위에서 코멘트 달아주셨는데, 저희도 AlertData 구조체를 굳이 만들지 않아도 된다고 생각하여 삭제하였습니다. 해당 수정은 [d456bd9](https://github.com/tasty-code/ios-contact-manager-ui/pull/34/commits/77fd2392458e0b9e31ae95bf4316bedc5d456bd9) 에서 확인하실 수 있습니다!

jungseungyeo commented on ContactManager/ContactManagerTests/ContactManagerTests.swift:
```diff
@@ -0,0 +1,62 @@
+import XCTest
+@testable import ContactManager
+
+final class ContactManagerTests: XCTestCase {
```
테스트 case가 다 실패하네요 ㅠ 이거 확인부탁드려요.

dpwns1234 replied:
[7c25f9](https://github.com/tasty-code/ios-contact-manager-ui/pull/34/commits/bd660a37c2f789dfc0934b7d91f5a39bae7c25f9) 커밋에서 수정하였습니다!

jungseungyeo commented on ContactManager/ContactManagerTests/ContactManagerTests.swift:
```diff
@@ -0,0 +1,62 @@
+  
+  func test_나이가_3자리_수_초과하면_false () {
+    let result = manager.ageValidate(1000)
+    XCTAssertFalse(result)
```
테스트 코드는 give, when then 패턴으로 주로 작성을 하고 있어요. 

위의 형태로 작성 부탁드려요.

dpwns1234 replied:
[7c25f9](https://github.com/tasty-code/ios-contact-manager-ui/pull/34/commits/bd660a37c2f789dfc0934b7d91f5a39bae7c25f9) 에서 수정하였습니다. give, when, then 패턴으로 작성하려고 하였지만, throw를 통해 validate를 작성하여서 when단계가 빠졌습니다. 괜찮을까요?

jungseungyeo commented on ContactManager/ContactManager/Controller/AddContactViewController.swift:
```diff
@@ -0,0 +1,88 @@
+      dismiss(animated: true)
+    } catch {
+      guard let errorType = error as? ValidationError else {
+        print(error.localizedDescription)
```
여기서도 에러가 발생할 수있는 코드네요. 여기서도 alert을 추가하면 더욱 안전한 코드가 될 거 같은데요.

AlertViewController을 활용해서 부탁드립니다.

dpwns1234 replied:
ValidationError에서 LocalizedError을 채택하여 catch문의 error를 형변환하지 않아도 되어서 다른 에러가 발생하지 않도록 구현하였습니다. 해당 수정은 https://github.com/tasty-code/ios-contact-manager-ui/pull/34/commits/1c7b48b7ee3f8aca162c3a314469e85d9c15f64a 커밋에서 반영되었습니다!

jungseungyeo commented on ContactManager/ContactManager/Controller/ContactManagerViewController.swift:
```diff
@@ -21,28 +20,34 @@ final class ContactManagerViewController: UIViewController, UITableViewDelegate
     contactTableView.dataSource = self
     
-    loadData()
+    NotificationCenter.default.addObserver(self, selector: #selector(reload), name: NSNotification.Name("AddContact"), object: nil)
```
NotificationCenter를 사용하면 안됩니다. 이 방법 말고 의존성 주입 또는 delegate패턴으로 이벤트를 전달해주세요

dpwns1234 replied:
넵! ContactChangedDelegate 를 사용해 delegate 패턴으로 수정하였습니다 https://github.com/tasty-code/ios-contact-manager-ui/pull/34/commits/892763eac45f0e1ec3c7c28347a6ffb542dcb0b9 커밋에서 반영했습니다!

jungseungyeo commented on ContactManager/ContactManager/Model/ContactManager.swift:
```diff
@@ -0,0 +1,67 @@
+    let data = Contact(name: name, age: age, phone: phone)
+    contacts.append(data)
+    NotificationCenter.default.post(name: NSNotification.Name("AddContact"),
+                                    object: self)
```
NotificationCenter말고 다른방법으로 해주세요. NotificationCenter은 브로드 캐스팅 방식으로 전역적으로 의존성이 발생할수있어요.

dpwns1234 replied:
넵 위에서 말해주신 대로 delegate 패턴으로 변경했습니다.

jungseungyeo commented on ContactManager/ContactManager/Model/ContactManager.swift:
```diff
@@ -0,0 +1,67 @@
+    guard let name = name?.components(separatedBy: " ").joined() else { throw ValidationError.nameValidationError }
+    if name.isEmpty { throw ValidationError.nameValidationError }
+    
+    return name
```
이거 객체보다는 extesnion으로 기능을 확장하면 어떤가요? 객체로 존재하는 이유가 있을까요?

dpwns1234 replied:
알려주신 대로 validation을 String을 확장하여 구현하였습니다. 해당 코드 수정은 https://github.com/tasty-code/ios-contact-manager-ui/pull/34/commits/b1885c107e0e788d772e4f3f3d6b22b29b8ff5af 커밋에서 반영했습니다!

jungseungyeo reviewed (changes requested):
UML이 잘 작성되었다고 생각하시는지 궁금합니다! 잘못된 점이나 보충해야할 부분이 있을까요~?

`수정이 필요한 부분이 보여요 화살표는 의존성이 있는쪽으로 되어야합니다. UML을 보시면 AddContactViewController이 ContactManagerViewController쪽으로 화살표가 되어 있는데 반대로 되어야해요. ContactManagerViewController -> AddContactViewController를 의존하고 있는 그림을 보여주여야합니다. 이와 맞찬가지로 AlertViewController를 가리키는 것도 동일합니다. `

고민한 점 4번에서 manager를 전달할 때 Singleton 패턴은 이전 프로젝트로 구현해 본 경험이 있어 다른 방식을 찾던 중 의존성 주입을 하는 형태를 알게되어 해당 방식으로 구현했습니다. 작성한 코드가 의존성 주입을 올바르게 처리한 방법인지 명확하지 않아 확인 부탁드립니다!

`의존성 주입이 되었다기보다는 그냥 의존성을 가지고 있는 형태이네요. 의존성 주입을 통해 테스트가 가능한 구조로 설계가 되어야하는데 바로 접근해서 사용하여 테스트가 불가능한 상태입니다. 의존성 주입을 하기 위해서는 protocol을 AddContactViewController의 채택하고 주입해주시면 좋을거같아요.`

중복되는 코드를 줄이고자 AlertViewController를 만들어 구현했습니다. 이전 PR에서 동일한 문제 해결방식을 물어보았는데 Factory pattern을 활용해보라는 답변을 받았다고 해서 고민해보았습니다.

`여러 방법이 있겠지만 지금 구조에서는 많은 인터페이스를 만들어주는 것이 좋아보이고 static하게 접근해서 사용하는 방법이 좋아보여요. static func으로 구성을하고 필요한 시점에 접근해서 사용하는 방식으로 해결하시면 좋을거같네요.`

AlertViewController의 showAlert()메서드를 호출할 때 너무 많은 매개변수를 받는 것 같아, AlertData 구조체를 만들어 showAlert(AlertData)로 전달하도록 구현했습니다. 이렇게 리팩터하는 것이 맞는지, 매개변수의 개수를 몇 개를 넘어서면 안되는지에 대한 기준이 있는지 궁금합니다.

`매개변수의 갯수는 제한이 없죠. 필요하면 쓰고 필요없다면 안쓰는게 맞겠죠. 하지만 이러한 것은 기준이 필요합니다. 단순히 몇개를 정하기 보다는 왜 무엇이 필요한지 고민이 필요해요. AlertData를 만들어서 사용하면 무엇이 좋을까요? 데이터를 다른곳에서 수정후 다시 사용하기 용이하겠죠? 하지만 이러한 AlertData는 누가 가지고 있을 필요가 없고, 수정도 필요한 부분이 보이지 않아요. 이러한 경우에는 저는 좋지 않다고 판단하고 있어요. 내부적으로 재사용이 가능한 구조를 만들기 위해서 주입이 필요한 configure를 만드는것은 큰 의미가 없는거 같더라구요. 현재 구조에서는 저는 큰 도움이 되는 AlertData 같지는 않습니다. 그렇다고 꼭 없애도 달라는 것은 아니라 기준을 잡고 진행해주세요 :) `

rarlala committed:
refactor: 핸드폰 - 추가 함수 별도 Utility 폴더로 분리

dpwns1234 committed:
refactor: AlertData 구조체 삭제 및 showAlert을 static 함수로 바꿈

- 불필요한 alertData 삭제
- 재사용성을 위해 static 함수로 변환

dpwns1234 committed:
test: 동작하지 않는 테스트 코드 수정

- given, when, then 형식 기반으로 작성

dpwns1234 committed:
refactor: ValidationError enum의 LocalizedError 채택

- ValidationError로 형변환 하지 않고 error에서 localizedDescription 으로 바로 접근 가능

dpwns1234 committed:
refactor: validate를 string extension으로 확장

dpwns1234 committed:
refactor: ContactManager 제거 및 NotificationCenter 제거

- ContactManager 제거 후 ContactMangerVC의 Controller 역할 재배정
- NotificationCenter 제거 후 delegate 패턴으로 변경

dpwns1234 committed:
refactor: AlertController show 메서드 오버로딩

dpwns1234 commented:
> UML이 잘 작성되었다고 생각하시는지 궁금합니다! 잘못된 점이나 보충해야할 부분이 있을까요~?
> 
> `수정이 필요한 부분이 보여요 화살표는 의존성이 있는쪽으로 되어야합니다. UML을 보시면 AddContactViewController이 ContactManagerViewController쪽으로 화살표가 되어 있는데 반대로 되어야해요. ContactManagerViewController -> AddContactViewController를 의존하고 있는 그림을 보여주여야합니다. 이와 맞찬가지로 AlertViewController를 가리키는 것도 동일합니다. `

해당 답변에 추가 질문이 있어 커멘트 답니다!
<img width="520" alt="스크린샷 2023-10-17 오후 1 36 48" src="https://github.com/tasty-code/ios-contact-manager-ui/assets/52391722/fd021be6-83d3-4d3a-b55f-44d9d16df24c">

https://www.nextree.co.kr/p6753/

저희가 학습한 링크에는 위 그림과 같이 참조하고 있는 곳에 화살표를 그리고 있습니다.
address가 user에서 참조될 때 화살표의 방향이 user -> address 로 표현되어 저희의 UML도
AddContactViewController가 ContactManager를 참조하고 있기에 해당 방식과 같이 표현했습니다.
(AddContactViewController -> ContactManager)

린생은 화살표가 의존성이 있는 쪽으로 되어야 한다고 하셨는데 이 부분에 대해 잘 이해가 되지 않는 것 같습니다..!

dpwns1234 committed:
refactor: 함수의 순서 정렬 및 final 키워드 추가

jungseungyeo commented on ContactManager/ContactManager/Utility/FormatingPhone.swift:
```diff
@@ -0,0 +1,32 @@
+
+import Foundation
+
+extension String {
```
너무 좋아요 ㅎㅎ 사용해보니깐 어떤가요? 더욱 깔끔해진거같네요 ㅎㅎ

jungseungyeo commented on ContactManager/ContactManagerTests/ContactManagerTests.swift:
```diff
@@ -11,52 +11,54 @@ import XCTest
+    // then
+    XCTAssertThrowsError(try manager.addContact(nameText: input, ageText: "22", phoneText: "010-4234-6212")) { error in
+      XCTAssertEqual(error as! ValidationError, ValidationError.nameValidationError)
+    }
```
테스트는 일반적으로 Equal의 형태가 되어야 됩니다. 에러를 발생했는지로는 테스트를 하시면 정확한 테스트가 불가능해요

jungseungyeo commented on ContactManager/ContactManager/Controller/AddContactViewController.swift:
```diff
@@ -22,25 +22,25 @@ class AddContactViewController: UIViewController {
+                             destructiveButtonTitle: "예",
+                             destructiveAction: {
+      self.dismiss(animated: true)
+    })
```
static 인터페이스를 만드니깐 좀 더 사용성이 좋아진거같네요 ㅎㅎ

jungseungyeo reviewed (approved):
고생했습니다. 

좀 더 수정이 필요한부분이 보이지만 리뷰에서 남겨두었습니다.
다음 step 진행 부탁드립니다.
